src/v/container/include/container/fragmented_vector.h

```diff
@@ -320,15 +426,24 @@ class fragmented_vector {
 
 private:
     void maybe_add_capacity() {
```

Comment:
So for chunked vector this relies on the first fragment going up to exactly the right amount without our explicit control?

Can't we just manage memory like we did before, with `_capacity` but with an additional condition when there is 1 frag && is_chunked_vector?

This would "solve" the problem mentioned in shrink_to_fit, and the over-allocation in `reserve()` which both can waste a fair amount of memory: having `reserve` is a really nice feature of chunked_vector.

I'm thinking something along these lines:

```
void add_capacity() {
        if constexpr (is_chunked_vector) {
            if (
              _frags.size() == 1 && _frags.back().capacity() < elems_per_frag) {
                auto new_cap = std::min(
                  elems_per_frag, _frags.back().capacity() * 2);
                _capacity += new_cap - _frags.back().capacity();
                _frags.back().reserve(new_cap);
                return;
            } else if (_frags.empty()) {
                _capacity = 10;
                _frags.emplace_back().reserve(_capacity);
                return;
            }
        }
        _frags.emplace_back().reserve(elems_per_frag);
        _capacity += elems_per_frag;
    }
```

10 should be replaced by a `constexpr` initial capacity which is like how many elements fit in 32 bytes or something like that since doing many allocations for the first elements, when they are small is quite wasteful since the memory savings is slow as there is already a significant fixed overhead in the structure itself.

This benchmarks about the same as the existing one (using Lifo and Fifo tests only).



```
if (is_chunked_vector &&

Reply:
> So for chunked vector this relies on the first fragment going up to exactly the right amount without our explicit control?

Yupp. There are tests that the "right thing" happens, but agreed anything seems fragile that isn't re-creating the inner container to not be std::vector but something custom. That feels like a risker change that would probably take more care and time than I can devote at the moment.

> but with other types of mutation it seems hard to predict exactly what the vector will do (plus implementation may due some rounding/adjusting as we do ).

Yeah in general there is not garentees by the stdlib, but practically we can test (or just not use the stdlib).

I've adapted this method, I think it is less error prone.

Reply:
> This would "solve" the problem mentioned in shrink_to_fit

Only when calling shrink_to_fit on `_frags.size() == 1`, otherwise if you have 5 fragments allocated, call shrink to fit on the last and then push back, you could get an overalloc.

Reply:
Fun fact, `std::vector::reserve();` seems to round up to a power of 2 automatically. At least in debug mode. They are using allocate_at_least.

Reply:
@rockwotj wrote:

> Yupp. There are tests that the "right thing" happens, but agreed anything seems fragile that isn't re-creating the inner container to not be std::vector but something custom.

Right, I don't want you to re-create the inner vector, but I think you can handle it by managing capacity along the lines I suggested. Current push looks good on this aspect  modulo comments.